packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -0,0 +1,333 @@
+      ) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+};
+
+const findSafe = (safe: SafeSnapSafe, chainId: number, ogAddress: string): boolean => {
+  return (
+    safe.network === chainId.toString() && ethersUtils.getAddress(safe.umaAddress) === ethersUtils.getAddress(ogAddress)
+  );
+};
+
+export const verifyProposal = async (
+  transaction: TransactionsProposedEvent,
+  params: MonitoringParams
+): Promise<VerificationResponse> => {
+  const ipfsHash = tryHexToUtf8String(transaction.args.explanation);
+  // tryHexToUtf8String returns the input if it is not decoding to UTF-8 string.
+  if (ipfsHash === transaction.args.explanation) {
+    return { verified: false, error: `Could not decode explanation ${transaction.args.explanation}` };
+  }
+
+  // Get proposal data from GraphQL.
+  const graphqlData = await getGraphqlData(ipfsHash, params.graphqlEndpoint);
+  if (graphqlData instanceof Error) {
+    return { verified: false, error: `GraphQL request failed with error ${graphqlData.message}` };
+  }
+
+  // Verify that the proposal exists and is unique.
+  if (graphqlData.proposals.length === 0) {
+    return { verified: false, error: `No proposal found for IPFS hash ${ipfsHash}` };
+  } else if (graphqlData.proposals.length > 1) {
+    return { verified: false, error: `Duplicate proposals found for IPFS hash ${ipfsHash}` };
+  }
+
+  // Verify proposal type.
+  const proposal = graphqlData.proposals[0];
+  if (proposal.type !== "single-choice" && proposal.type !== "basic") {
+    return { verified: false, error: `Proposal type ${proposal.type} is not supported` };
+  }
+  if (proposal.type === "basic") {
+    // Assert that the basic proposal has expected choices.
+    assert(proposal.choices.length === 3, "Basic proposal must have three choices");
+    assert(proposal.choices[0] === "For", "Basic proposal must have 'For' as the first choice");
+    assert(proposal.choices[1] === "Against", "Basic proposal must have 'Against' as the second choice");
+    assert(proposal.choices[2] === "Abstain", "Basic proposal must have 'Abstain' as the third choice");
+  }
+
+  // Verify that the proposal is in the closed state.
+  if (proposal.state !== "closed") {
+    return { verified: false, error: "Proposal not in closed state" };
+  }
+
+  // Verify that the proposal voting period has ended.
+  if (transaction.args.proposalTime.toNumber() < proposal.end) {
+    return { verified: false, error: "Proposal voting period has not ended" };
+  }
+
+  // Verify quorum.
+  if (proposal.scores_total !== null && proposal.scores_total < proposal.quorum) {
+    return { verified: false, error: `Proposal did not meet Snapshot quorum of ${proposal.quorum}` };
+  }
+
+  // Verify proposal choices and scores.
+  const approvalIndex = proposal.choices.findIndex((choice) =>
+    params.approvalChoices.map((approvalChoice) => approvalChoice.toLowerCase()).includes(choice.toLowerCase())
+  );
```

Comment:
Is this always finding the approvalIndex? It seems that it finds the first occurrency but I'm not sure if it's necessarily the approved option.

Reply:
If none of choices matches known approval labels this should return `-1` that returns the proposal as unverified.

But its a good call to also check if multiple choices matched. E.g. if there were both `Yes` and `For` in the same proposal, it means misconfiguration that we should escalate. I'll fix this to count all occurrences of known approval labels and make sure there is one and only one match.